File: src/main/perl/Mp4DefaultTrack.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Fcntl qw(:seek);

# -----------------------------------------
# MP4 language decode (ISO-639-2/T)
# -----------------------------------------
sub decode_lang {
    my ($packed) = @_;
    return undef if !$packed || $packed == 0;
    my $a = chr((($packed >> 10) & 0x1F) + 96);
    my $b = chr((($packed >> 5)  & 0x1F) + 96);
    my $c = chr(( $packed        & 0x1F) + 96);
    return "$a$b$c";
}

# -----------------------------------------
# Parse MP4 structure
# -----------------------------------------
sub parse_mp4 {
    my ($fh) = @_;
    my $file_len = -s $fh;
    my @tracks;

    my $p = 0;
    sysseek($fh, 0, SEEK_SET);
    
    while ($p + 8 <= $file_len) {
        my $hdr;
        last if sysread($fh, $hdr, 8) != 8;
        my ($boxsize, $type) = unpack("Na4", $hdr);

        if ($boxsize == 1) {
            last if sysread($fh, $hdr, 8) != 8;
            my ($hi, $lo) = unpack("NN", $hdr);
            $boxsize = $hi * 4294967296 + $lo;
        } elsif ($boxsize == 0) {
            $boxsize = $file_len - $p;
        }
        last if $boxsize < 8;

        if ($type eq "moov") {
            parse_moov($fh, $p, $boxsize, \@tracks);
        }

        $p += $boxsize;
        sysseek($fh, $p, SEEK_SET);
    }
    return @tracks;
}

sub parse_moov {
    my ($fh, $start, $size, $tracks) = @_;
    my $end = $start + $size;
    my $p = $start + 8;

    sysseek($fh, $p, SEEK_SET);
    while ($p + 8 <= $end) {
        my $hdr;
        last if sysread($fh, $hdr, 8) != 8;
        my ($boxsize, $type) = unpack("Na4", $hdr);
        
        if ($boxsize == 1) {
            sysread($fh, $hdr, 8);
            my ($hi, $lo) = unpack("NN", $hdr);
            $boxsize = $hi * 4294967296 + $lo;
        } elsif ($boxsize == 0) {
            $boxsize = $end - $p;
        }
        last if $boxsize < 8;

        if ($type eq "trak") {
            my $t = parse_trak($fh, $p, $boxsize);
            push @$tracks, $t if $t;
        }

        $p += $boxsize;
        sysseek($fh, $p, SEEK_SET);
    }
}

sub parse_trak {
    my ($fh, $start, $size) = @_;
    my $end = $start + $size;
    my $p = $start + 8;

    my %t = (
        trackId     => 0,
        type        => "unknown",
        language    => undef,
        tkhdOffset  => undef,
        stsdOffset  => undef,
        defaultFlag => 0,
        forcedFlag  => 0,
    );

    sysseek($fh, $p, SEEK_SET);
    while ($p + 8 <= $end) {
        my $hdr;
        last if sysread($fh, $hdr, 8) != 8;
        my ($boxsize, $type) = unpack("Na4", $hdr);
        
        if ($boxsize == 1) {
            sysread($fh, $hdr, 8);
            my ($hi, $lo) = unpack("NN", $hdr);
            $boxsize = $hi * 4294967296 + $lo;
        } elsif ($boxsize == 0) {
            $boxsize = $end - $p;
        }
        last if $boxsize < 8;

        if ($type eq "tkhd") {
            $t{tkhdOffset} = $p + 8;
            my $buf;
            # Read enough to cover version, flags, and Track ID in one call
            if (sysread($fh, $buf, 24) == 24) {
                my $ver = unpack("C", substr($buf, 0, 1));
                my $flags = unpack("N", "\x00" . substr($buf, 1, 3));
                $t{defaultFlag} = ($flags & 1);
                
                my $id_offset = ($ver == 1) ? 16 : 8;
                $t{trackId} = unpack("N", substr($buf, 4 + $id_offset, 4));
            }
        } elsif ($type eq "mdia") {
            parse_mdia($fh, $p, $boxsize, \%t);
        }

        $p += $boxsize;
        sysseek($fh, $p, SEEK_SET);
    }
    return $t{trackId} ? \%t : undef;
}

sub parse_mdia {
    my ($fh, $start, $size, $t) = @_;
    my $end = $start + $size;
    my $p = $start + 8;

    sysseek($fh, $p, SEEK_SET);
    while ($p + 8 <= $end) {
        my $hdr;
        last if sysread($fh, $hdr, 8) != 8;
        my ($boxsize, $type) = unpack("Na4", $hdr);
        
        if ($boxsize == 1) {
            sysread($fh, $hdr, 8);
            my ($hi, $lo) = unpack("NN", $hdr);
            $boxsize = $hi * 4294967296 + $lo;
        } elsif ($boxsize == 0) {
            $boxsize = $end - $p;
        }
        last if $boxsize < 8;

        if ($type eq "mdhd") {
            $t->{mdhdOffset} = $p + 8;
            my $buf;
            sysread($fh, $buf, 1);
            my $ver = unpack("C", $buf);
            my $skip = ($ver == 1) ? 28 : 16;
            
            sysseek($fh, $p + 8 + 4 + $skip, SEEK_SET);
            if (sysread($fh, $buf, 2) == 2) {
                $t->{language} = decode_lang(unpack("n", $buf));
            }
        } elsif ($type eq "hdlr") {
            sysseek($fh, $p + 16, SEEK_SET);
            my $sub;
            if (sysread($fh, $sub, 4) == 4) {
                if    ($sub eq "vide") { $t->{type} = "video"; }
                elsif ($sub eq "soun") { $t->{type} = "audio"; }
                elsif ($sub eq "subt" || $sub eq "sbtl" || $sub eq "text") { $t->{type} = "subtitle"; }
                else                   { $t->{type} = $sub; }
            }
        } elsif ($type eq "minf") {
            parse_minf($fh, $p, $boxsize, $t);
        }

        $p += $boxsize;
        sysseek($fh, $p, SEEK_SET);
    }
}

sub parse_minf {
    my ($fh, $start, $size, $t) = @_;
    my $end = $start + $size;
    my $p = $start + 8;

    sysseek($fh, $p, SEEK_SET);
    while ($p + 8 <= $end) {
        my $hdr;
        last if sysread($fh, $hdr, 8) != 8;
        my ($boxsize, $type) = unpack("Na4", $hdr);
        
        if ($boxsize == 1) {
            sysread($fh, $hdr, 8);
            my ($hi, $lo) = unpack("NN", $hdr);
            $boxsize = $hi * 4294967296 + $lo;
        } elsif ($boxsize == 0) {
            $boxsize = $end - $p;
        }
        last if $boxsize < 8;

        if ($type eq "stbl") {
            parse_stbl($fh, $p, $boxsize, $t);
        }

        $p += $boxsize;
        sysseek($fh, $p, SEEK_SET);
    }
}

sub parse_stbl {
    my ($fh, $start, $size, $t) = @_;
    my $end = $start + $size;
    my $p = $start + 8;

    sysseek($fh, $p, SEEK_SET);
    while ($p + 8 <= $end) {
        my $hdr;
        last if sysread($fh, $hdr, 8) != 8;
        my ($boxsize, $type) = unpack("Na4", $hdr);
        
        if ($boxsize == 1) {
            sysread($fh, $hdr, 8);
            my ($hi, $lo) = unpack("NN", $hdr);
            $boxsize = $hi * 4294967296 + $lo;
        } elsif ($boxsize == 0) {
            $boxsize = $end - $p;
        }
        last if $boxsize < 8;

        if ($type eq "stsd") {
            $t->{stsdOffset} = $p + 8;
            sysseek($fh, $p + 20, SEEK_SET);
            my $sampleType;
            if (sysread($fh, $sampleType, 4) == 4) {
                $t->{forcedFlag} = ($sampleType =~ /fcd /i) ? 1 : 0;
            }
        }

        $p += $boxsize;
        sysseek($fh, $p, SEEK_SET);
    }
}

# -----------------------------------------
# Patchers (Using existing active FileHandle)
# -----------------------------------------
sub patch_default {
    my ($fh, $offset, $set) = @_;
    sysseek($fh, $offset + 1, SEEK_SET) or die $!;
    
    my $buf;
    sysread($fh, $buf, 3) == 3 or die "short read";
    my ($b1, $b2, $b3) = unpack("C3", $buf);
    my $flags = ($b1 << 16) | ($b2 << 8) | $b3;

    if ($set) { $flags |= 1; }
    else      { $flags &= ~1; }

    my $out = pack("C3", ($flags >> 16) & 255, ($flags >> 8) & 255, $flags & 255);
    sysseek($fh, $offset + 1, SEEK_SET);
    syswrite($fh, $out) == 3 or die "short write";
}

sub patch_forced {
    my ($fh, $stsd, $set) = @_;
    return unless $stsd && $set;
    
    sysseek($fh, $stsd + 12, SEEK_SET);
    syswrite($fh, "fcd ") == 4 or die "short write";
}

# -----------------------------------------
# JSON printing
# -----------------------------------------
sub print_json_list {
    my (@tracks) = @_;
    print "[\n";
    for my $i (0 .. $#tracks) {
        my $t = $tracks[$i];
        my $lang  = defined($t->{language}) ? "\"$t->{language}\"" : "null";
        my $def   = $t->{defaultFlag} ? "true" : "false";
        my $fcd   = $t->{forcedFlag} ? "true" : "false";
        my $comma = ($i < $#tracks) ? "," : "";
        print "\t{\"id\": $t->{trackId}, \"type\": \"$t->{type}\", \"lang\": $lang, \"default\": $def, \"forced\": $fcd}$comma\n";
    }
    print "]\n";
}

# -----------------------------------------
# CLI
# -----------------------------------------
sub main {
    my ($cmd, $file, $id, $flag) = @ARGV;
    if (!$cmd || !$file) {
        die "Usage:\n" .
            "  perl mp4track.pl list <file>\n" .
            "  perl mp4track.pl set <file> <id> <default|forced>\n" .
            "  perl mp4track.pl unset <file> <id> <default|forced>\n";
    }

    # Open file ONCE. Read-only for listing, read-write for patching.
    my $mode = ($cmd eq "list") ? "<:raw" : "+<:raw";
    open(my $fh, $mode, $file) or die "Cannot open $file: $!\n";

    if ($cmd eq "list") {
        my @tracks = parse_mp4($fh);
        print_json_list(@tracks);
        close($fh);
        exit 0;
    }

    if ($cmd eq "set" || $cmd eq "unset") {
        die "Missing args: need id + flag\n" unless $id && $flag;
        
        my @tracks = parse_mp4($fh);
        my $tid = int($id);
        my $found = 0;

        for my $t (@tracks) {
            next unless $t->{trackId} == $tid;
            $found = 1;

            if ($flag eq "default") {
                patch_default($fh, $t->{tkhdOffset}, $cmd eq "set");
            } elsif ($flag eq "forced") {
                patch_forced($fh, $t->{stsdOffset}, $cmd eq "set");
            } else {
                close($fh);
                die "Unknown flag: $flag\n";
            }
            last;
        }
        
        close($fh);
        die "Track not found\n" unless $found;
        exit 0;
    }

    die "Unknown command: $cmd\n";
}

main();

File: src/main/python/Mp4DefaultTrack.py
#!/usr/bin/env python3
import sys
import mmap
import json

# ----------------------------
#   Constants
# ----------------------------
# Comparing raw bytes is dramatically faster than string decoding
B_MOOV = b"moov"
B_TRAK = b"trak"
B_TKHD = b"tkhd"
B_MDIA = b"mdia"
B_MDHD = b"mdhd"
B_HDLR = b"hdlr"
B_MINF = b"minf"
B_STBL = b"stbl"
B_STSD = b"stsd"
B_VIDE = b"vide"
B_SOUN = b"soun"
B_SUBT = b"subt"
B_SBTL = b"sbtl"
B_TEXT = b"text"
B_FCD_ = b"fcd "

# ----------------------------
#   Helpers
# ----------------------------

def decode_language(bits):
    if bits == 0:
        return None
    return "".join(chr(((bits >> shift) & 0x1F) + 0x60) for shift in (10, 5, 0))

def iter_boxes(mm, start, end):
    """Yields (payload_offset, header_length, total_box_size, box_type_bytes)"""
    pos = start
    while pos + 8 <= end:
        # Native int.from_bytes is significantly faster than struct.unpack
        s = int.from_bytes(mm[pos:pos+4], 'big')
        typ = mm[pos+4:pos+8]
        box_size = s
        hdr_len = 8

        if s == 1:
            if pos + 16 > end:
                break
            box_size = int.from_bytes(mm[pos+8:pos+16], 'big')
            hdr_len = 16
        elif s == 0:
            box_size = end - pos

        if box_size < 8:
            break

        yield pos, hdr_len, box_size, typ
        pos += box_size

# ----------------------------
# TrackInfo
# ----------------------------

class Track:
    # __slots__ strictly bounds the memory footprint of the objects
    __slots__ = ['tkhd_offset', 'stsd_offset', 'mdhd_offset', 'track_id', 'default', 'forced', 'type', 'lang']

    def __init__(self):
        self.tkhd_offset = None
        self.stsd_offset = None
        self.mdhd_offset = None
        self.track_id = 0
        self.default = False
        self.forced = False
        self.type = None
        self.lang = None

# ----------------------------
# Main parsing
# ----------------------------

def parse_mp4(mm):
    tracks = []
    for pos, hdr, size, typ in iter_boxes(mm, 0, len(mm)):
        if typ == B_MOOV:
            parse_moov(mm, pos + hdr, pos + size, tracks)
    return tracks

def parse_moov(mm, start, end, tracks):
    for pos, hdr, size, typ in iter_boxes(mm, start, end):
        if typ == B_TRAK:
            t = parse_trak(mm, pos + hdr, pos + size)
            if t and t.track_id != 0:
                tracks.append(t)

def parse_trak(mm, start, end):
    t = Track()
    for pos, hdr, size, typ in iter_boxes(mm, start, end):
        if typ == B_TKHD:
            t.tkhd_offset = pos + hdr
            version = mm[t.tkhd_offset]
            flags = int.from_bytes(mm[t.tkhd_offset+1:t.tkhd_offset+4], 'big')
            t.default = bool(flags & 1)

            skip = 16 if version == 1 else 8
            id_pos = t.tkhd_offset + 4 + skip
            t.track_id = int.from_bytes(mm[id_pos:id_pos+4], 'big')

        elif typ == B_MDIA:
            parse_mdia(mm, pos + hdr, pos + size, t)
    return t

def parse_mdia(mm, start, end, t):
    for pos, hdr, size, typ in iter_boxes(mm, start, end):
        if typ == B_MDHD:
            t.mdhd_offset = pos + hdr
            version = mm[t.mdhd_offset]
            skip = 28 if version == 1 else 16
            lang_pos = t.mdhd_offset + 4 + skip
            lang_bits = int.from_bytes(mm[lang_pos:lang_pos+2], 'big')
            t.lang = decode_language(lang_bits)

        elif typ == B_HDLR:
            subtype = mm[pos + hdr + 8 : pos + hdr + 12]
            if subtype == B_VIDE:
                t.type = "video"
            elif subtype == B_SOUN:
                t.type = "audio"
            elif subtype in (B_SUBT, B_SBTL, B_TEXT):
                t.type = "subtitle"
            else:
                t.type = subtype.decode("ascii", errors="ignore")

        elif typ == B_MINF:
            parse_minf(mm, pos + hdr, pos + size, t)

def parse_minf(mm, start, end, t):
    for pos, hdr, size, typ in iter_boxes(mm, start, end):
        if typ == B_STBL:
            parse_stbl(mm, pos + hdr, pos + size, t)

def parse_stbl(mm, start, end, t):
    for pos, hdr, size, typ in iter_boxes(mm, start, end):
        if typ == B_STSD:
            t.stsd_offset = pos + hdr
            entry_start = t.stsd_offset + 8
            if entry_start + 8 <= pos + size:
                sample_type = mm[entry_start+4:entry_start+8].lower()
                t.forced = b"fcd" in sample_type

# ----------------------------
# Write operations
# ----------------------------

def patch_default_flag(mm, offset, value):
    flags = int.from_bytes(mm[offset+1:offset+4], 'big')
    if value:
        flags |= 1
    else:
        flags &= ~1
    mm[offset+1:offset+4] = flags.to_bytes(3, 'big')

def patch_forced_flag(mm, offset, value):
    # If unsetting, original string type is unknown, ignore safely
    if value:
        mm[offset+12:offset+16] = B_FCD_

# ----------------------------
# Commands
# ----------------------------

def cmd_list(path):
    with open(path, "rb") as f:
        with mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as mm:
            tracks = parse_mp4(mm)

    out = []
    for t in tracks:
        out.append({
            "id": t.track_id,
            "type": t.type,
            "lang": t.lang,
            "default": t.default,
            "forced": t.forced
        })

    # Mirrors original JSON spacing output but heavily optimized
    print("[")
    for i, obj in enumerate(out):
        line = json.dumps(obj)
        print(f"\t{line}{',' if i < len(out)-1 else ''}")
    print("]")


def cmd_setunset(path, track_id, flag, value):
    # Open once in r+b mode and map with ACCESS_WRITE to modify the file directly
    with open(path, "r+b") as f:
        with mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_WRITE) as mm:
            tracks = parse_mp4(mm)
            for t in tracks:
                if t.track_id == track_id:
                    if flag == "default":
                        patch_default_flag(mm, t.tkhd_offset, value)
                    elif flag == "forced":
                        patch_forced_flag(mm, t.stsd_offset, value)
                    else:
                        sys.exit("Unknown flag.")
                    return
    print("Track not found.")

# ----------------------------
# Entry point
# ----------------------------

if __name__ == "__main__":
    if len(sys.argv) < 3:
        sys.exit("Usage: mp4track.py <list|set|unset> <file> [trackId] [default|forced]")

    cmd = sys.argv[1]
    path = sys.argv[2]

    try:
        if cmd == "list":
            cmd_list(path)
            sys.exit(0)

        if cmd in ("set", "unset"):
            if len(sys.argv) < 5:
                sys.exit("Missing args: set/unset <file> <trackId> <default|forced>")

            tid = int(sys.argv[3])
            flag = sys.argv[4]
            val = (cmd == "set")
            cmd_setunset(path, tid, flag, val)
            sys.exit(0)

        sys.exit("Unknown command.")
    except FileNotFoundError:
        sys.exit("File not found.")
